# EV Data wrangling consistency check and merging

## Table of content

1. Importing libraries

2. Importing data

3. Data wrangling

        3.1 Dropping columns
        3.2 Renaming columns
        3.3 Checking data types

4. Data consistency checks

        4.1 Missing values
        4.2 Duplicates    

5. Subsetting dataframes

        5.1 Subsetting income dataframes
        5.2. Population dataframe        

6. Merging dataframes

        6.1 Merging car_new and car_stock dataframes
        6.2 Merging car_merged and gdp dataframes
        6.3 Merging car_gdp and income dataframes
        6.4 Merging car_gdp_inc and pop_avg dataframes

7. Deriving new variables

        7.1 Engine_type
        7.2 Country
        7.3 Age
        7.4 Income indicator
        7.5 Coutry grouping

8. Exporting dataframe

# 1. Importing libraries

In [178]:
# Importing libraries
import pandas as pd
import numpy as np
import os

# 2. Importing data

In [179]:
# Creating shortcut path to data
path_data = r'C:\Users\henri\Documents\Tonia\Learning\CarrrerFoundry\Data Immersion\Achievement 6\EV Analysis in EU\02 Data'

In [180]:
path_data

'C:\\Users\\henri\\Documents\\Tonia\\Learning\\CarrrerFoundry\\Data Immersion\\Achievement 6\\EV Analysis in EU\\02 Data'

### Importing New passenger cars registration dataset 

In [181]:
# Importing New passenger cars registration.csv
car_new = pd.read_csv(os.path.join(path_data, 'Original Data', 'New passenger cars registration.csv'))

In [182]:
car_new.head()

DATAFLOW        LAST UPDATE freq unit mot_nrg geo  \
0  ESTAT:ROAD_EQR_CARPDA(1.0)  30/03/23 23:00:00    A   NR     ALT  AL   
1  ESTAT:ROAD_EQR_CARPDA(1.0)  30/03/23 23:00:00    A   NR     ALT  AL   
2  ESTAT:ROAD_EQR_CARPDA(1.0)  30/03/23 23:00:00    A   NR     ALT  AL   
3  ESTAT:ROAD_EQR_CARPDA(1.0)  30/03/23 23:00:00    A   NR     ALT  AT   
4  ESTAT:ROAD_EQR_CARPDA(1.0)  30/03/23 23:00:00    A   NR     ALT  AT   

   TIME_PERIOD  OBS_VALUE OBS_FLAG  
0         2019     3757.0      NaN  
1         2020     4935.0      NaN  
2         2021     5703.0      NaN  
3         2013     1285.0      NaN  
4         2014     2074.0      NaN

In [183]:
car_new.shape

(3369, 9)

### Importing Stock passenger cars dataset

In [184]:
# Importing Stock passenger cars.csv
car_stock = pd.read_csv(os.path.join(path_data, 'Original Data', 'Stock passenger cars.csv'))

In [185]:
car_stock.head()

DATAFLOW        LAST UPDATE freq unit mot_nrg geo  \
0  ESTAT:ROAD_EQS_CARPDA(1.0)  30/03/23 23:00:00    A   NR     ALT  AL   
1  ESTAT:ROAD_EQS_CARPDA(1.0)  30/03/23 23:00:00    A   NR     ALT  AL   
2  ESTAT:ROAD_EQS_CARPDA(1.0)  30/03/23 23:00:00    A   NR     ALT  AL   
3  ESTAT:ROAD_EQS_CARPDA(1.0)  30/03/23 23:00:00    A   NR     ALT  AL   
4  ESTAT:ROAD_EQS_CARPDA(1.0)  30/03/23 23:00:00    A   NR     ALT  AL   

   TIME_PERIOD  OBS_VALUE OBS_FLAG  
0         2014     4373.0      NaN  
1         2015     7562.0      NaN  
2         2016    14544.0      NaN  
3         2017    21409.0      NaN  
4         2018    28994.0      NaN

In [186]:
car_stock.shape

(4751, 9)

### Importing GDP EU dataset

In [187]:
# Importing GDP EU.csv
gdp = pd.read_csv(os.path.join(path_data, 'Original Data', 'GDP EU.csv'))

In [188]:
gdp.head()

DATAFLOW        LAST UPDATE freq     unit  \
0  ESTAT:NAMA_10_GDP$DEFAULTVIEW(1.0)  26/04/23 23:00:00    A  CP_MEUR   
1  ESTAT:NAMA_10_GDP$DEFAULTVIEW(1.0)  26/04/23 23:00:00    A  CP_MEUR   
2  ESTAT:NAMA_10_GDP$DEFAULTVIEW(1.0)  26/04/23 23:00:00    A  CP_MEUR   
3  ESTAT:NAMA_10_GDP$DEFAULTVIEW(1.0)  26/04/23 23:00:00    A  CP_MEUR   
4  ESTAT:NAMA_10_GDP$DEFAULTVIEW(1.0)  26/04/23 23:00:00    A  CP_MEUR   

  na_item geo  TIME_PERIOD  OBS_VALUE OBS_FLAG  
0    B1GQ  AL         2013     9625.4      NaN  
1    B1GQ  AL         2014     9968.6      NaN  
2    B1GQ  AL         2015    10264.1      NaN  
3    B1GQ  AL         2016    10719.9      NaN  
4    B1GQ  AL         2017    11559.0      NaN

In [189]:
gdp.shape

(448, 9)

### Importing Mean and median income dataset

In [190]:
# Importing Mean and median income.csv
income = pd.read_csv(os.path.join(path_data, 'Original Data', 'Mean and median income.csv'))

In [191]:
income.head()

DATAFLOW        LAST UPDATE freq    age sex indic_il unit geo  \
0  ESTAT:ILC_DI03(1.0)  03/04/23 23:00:00    A  TOTAL   F    MED_E  EUR  AL   
1  ESTAT:ILC_DI03(1.0)  03/04/23 23:00:00    A  TOTAL   F    MED_E  EUR  AL   
2  ESTAT:ILC_DI03(1.0)  03/04/23 23:00:00    A  TOTAL   F    MED_E  EUR  AL   
3  ESTAT:ILC_DI03(1.0)  03/04/23 23:00:00    A  TOTAL   F    MED_E  EUR  AL   
4  ESTAT:ILC_DI03(1.0)  03/04/23 23:00:00    A  TOTAL   F    MED_E  EUR  AT   

   TIME_PERIOD  OBS_VALUE OBS_FLAG  
0         2017       1747      NaN  
1         2018       1972      NaN  
2         2019       2195      NaN  
3         2020       2482      NaN  
4         1995      13394      NaN

In [192]:
income.shape

(271872, 11)

### Importing Population dataset

In [193]:
# Importing Population.csv
pop = pd.read_csv(os.path.join(path_data, 'Original Data', 'Population.csv'))

In [194]:
pop.head()

DATAFLOW        LAST UPDATE freq indic_de geo  TIME_PERIOD  \
0  ESTAT:DEMO_GIND(1.0)  14/04/23 23:00:00    A      AVG  AD         1986   
1  ESTAT:DEMO_GIND(1.0)  14/04/23 23:00:00    A      AVG  AD         1987   
2  ESTAT:DEMO_GIND(1.0)  14/04/23 23:00:00    A      AVG  AD         1988   
3  ESTAT:DEMO_GIND(1.0)  14/04/23 23:00:00    A      AVG  AD         1989   
4  ESTAT:DEMO_GIND(1.0)  14/04/23 23:00:00    A      AVG  AD         1990   

   OBS_VALUE OBS_FLAG  
0    45785.0      NaN  
1    47714.0      NaN  
2    49490.0      NaN  
3    50708.0      NaN  
4    52697.0      NaN

In [195]:
pop.shape

(61645, 8)

# 3. Data wrangling

## 3.1 Dropping columns

In the above dataframes we do not need the following columns for the analysis:
    
    DATAFLOW - indicates the data name
    LAST UPDATE - indicates the last update
    freq - indicates the frequency of update of the data, which is annually
    unit
    OBS_FLAG, which is observation status flag and can mean e.g. estimates or break in data; this does not change the analysis, as the rows with these flags will not be removed

I will therefore remove them from all the dataframes.        

#### car_new dataframe

In [196]:
car_new = car_new.drop(columns = ['DATAFLOW', 'LAST UPDATE', 'freq', 'unit', 'OBS_FLAG'])

In [197]:
car_new.head()

mot_nrg geo  TIME_PERIOD  OBS_VALUE
0     ALT  AL         2019     3757.0
1     ALT  AL         2020     4935.0
2     ALT  AL         2021     5703.0
3     ALT  AT         2013     1285.0
4     ALT  AT         2014     2074.0

#### car_stock dataframe

In [198]:
car_stock = car_stock.drop(columns = ['DATAFLOW', 'LAST UPDATE', 'freq', 'unit', 'OBS_FLAG'])

In [199]:
car_stock.head()

mot_nrg geo  TIME_PERIOD  OBS_VALUE
0     ALT  AL         2014     4373.0
1     ALT  AL         2015     7562.0
2     ALT  AL         2016    14544.0
3     ALT  AL         2017    21409.0
4     ALT  AL         2018    28994.0

#### gdp dataframe

In the gdp dataframe, we should remove additionally column 'na_item'. This column shows National account indicator, which has been already set for 'B1GQ' (Gross domestic product at market prices)

In [200]:
gdp = gdp.drop(columns = ['DATAFLOW', 'LAST UPDATE', 'freq', 'unit', 'na_item', 'OBS_FLAG'])

In [201]:
gdp.head()

geo  TIME_PERIOD  OBS_VALUE
0  AL         2013     9625.4
1  AL         2014     9968.6
2  AL         2015    10264.1
3  AL         2016    10719.9
4  AL         2017    11559.0

#### income dataframe

In [202]:
income = income.drop(columns = ['DATAFLOW', 'LAST UPDATE', 'freq', 'OBS_FLAG'])

In [203]:
income.head()

age sex indic_il unit geo  TIME_PERIOD  OBS_VALUE
0  TOTAL   F    MED_E  EUR  AL         2017       1747
1  TOTAL   F    MED_E  EUR  AL         2018       1972
2  TOTAL   F    MED_E  EUR  AL         2019       2195
3  TOTAL   F    MED_E  EUR  AL         2020       2482
4  TOTAL   F    MED_E  EUR  AT         1995      13394

#### pop dataframe

In [204]:
pop = pop.drop(columns = ['DATAFLOW', 'LAST UPDATE', 'freq', 'OBS_FLAG'])

In [205]:
pop.head()

indic_de geo  TIME_PERIOD  OBS_VALUE
0      AVG  AD         1986    45785.0
1      AVG  AD         1987    47714.0
2      AVG  AD         1988    49490.0
3      AVG  AD         1989    50708.0
4      AVG  AD         1990    52697.0

## 3.2 Renaming columns

### car_new dataframe

The following column name should be changed:

    MOT_NRG - engine_type
    GEO - country
    TIME_PERIOD - year
    OBS_VALUE - new_car

In [206]:
# Renaming "mot_nrg" column in orders dataframe
car_new.rename(columns = {'mot_nrg' : 'engine_type'}, inplace = True)

In [207]:
# Renaming "geo" column in orders dataframe
car_new.rename(columns = {'geo' : 'country'}, inplace = True)

In [208]:
# Renaming "TIME_PERIOD" column in orders dataframe
car_new.rename(columns = {'TIME_PERIOD' : 'year'}, inplace = True)

In [209]:
# Renaming "OBS_VALUE" column in orders dataframe
car_new.rename(columns = {'OBS_VALUE' : 'new_car'}, inplace = True)

In [210]:
car_new.head()

engine_type country  year  new_car
0         ALT      AL  2019   3757.0
1         ALT      AL  2020   4935.0
2         ALT      AL  2021   5703.0
3         ALT      AT  2013   1285.0
4         ALT      AT  2014   2074.0

### car_stock dataframe

The following column name should be changed:

    MOT_NRG - engine_type
    GEO - country
    TIME_PERIOD - year
    OBS_VALUE - stock_car

In [211]:
# Renaming "mot_nrg" column in orders dataframe
car_stock.rename(columns = {'mot_nrg' : 'engine_type'}, inplace = True)

In [212]:
# Renaming "geo" column in orders dataframe
car_stock.rename(columns = {'geo' : 'country'}, inplace = True)

In [213]:
# Renaming "TIME_PERIOD" column in orders dataframe
car_stock.rename(columns = {'TIME_PERIOD' : 'year'}, inplace = True)

In [214]:
# Renaming "OBS_VALUE" column in orders dataframe
car_stock.rename(columns = {'OBS_VALUE' : 'stock_car'}, inplace = True)

In [215]:
car_stock.head()

engine_type country  year  stock_car
0         ALT      AL  2014     4373.0
1         ALT      AL  2015     7562.0
2         ALT      AL  2016    14544.0
3         ALT      AL  2017    21409.0
4         ALT      AL  2018    28994.0

### gdp dataframe

The following column name should be changed:

    GEO - country
    TIME_PERIOD - year
    OBS_VALUE - gdp_eur

In [216]:
# Renaming "geo" column in orders dataframe
gdp.rename(columns = {'geo' : 'country'}, inplace = True)

In [217]:
# Renaming "TIME_PERIOD" column in orders dataframe
gdp.rename(columns = {'TIME_PERIOD' : 'year'}, inplace = True)

In [218]:
# Renaming "OBS_VALUE" column in orders dataframe
gdp.rename(columns = {'OBS_VALUE' : 'gdp_eur'}, inplace = True)

In [219]:
gdp.head()

country  year  gdp_eur
0      AL  2013   9625.4
1      AL  2014   9968.6
2      AL  2015  10264.1
3      AL  2016  10719.9
4      AL  2017  11559.0

### income dataframe

The following column name should be changed:

    indic_il (Income and living conditions indicator) - income_indicator
    unit - currency
    GEO - country
    TIME_PERIOD - year
    OBS_VALUE - income

In [220]:
# Renaming "indic_il" column in orders dataframe
income.rename(columns = {'indic_il' : 'income_indicator'}, inplace = True)

In [221]:
# Renaming "unit" column in orders dataframe
income.rename(columns = {'unit' : 'currency'}, inplace = True)

In [222]:
# Renaming "geo" column in orders dataframe
income.rename(columns = {'geo' : 'country'}, inplace = True)

In [223]:
# Renaming "TIME_PERIOD" column in orders dataframe
income.rename(columns = {'TIME_PERIOD' : 'year'}, inplace = True)

In [224]:
# Renaming "OBS_VALUE" column in orders dataframe
income.rename(columns = {'OBS_VALUE' : 'income'}, inplace = True)

In [225]:
income.head()

age sex income_indicator currency country  year  income
0  TOTAL   F            MED_E      EUR      AL  2017    1747
1  TOTAL   F            MED_E      EUR      AL  2018    1972
2  TOTAL   F            MED_E      EUR      AL  2019    2195
3  TOTAL   F            MED_E      EUR      AL  2020    2482
4  TOTAL   F            MED_E      EUR      AT  1995   13394

### pop dataframe

The following column name should be changed:

    indic_de - demographic_indicator
    GEO - country
    TIME_PERIOD - year
    OBS_VALUE - population

In [226]:
# Renaming "indic_de" column in orders dataframe
pop.rename(columns = {'indic_de' : 'demographic_indicator'}, inplace = True)

In [227]:
# Renaming "geo" column in orders dataframe
pop.rename(columns = {'geo' : 'country'}, inplace = True)

In [228]:
# Renaming "TIME_PERIOD" column in orders dataframe
pop.rename(columns = {'TIME_PERIOD' : 'year'}, inplace = True)

In [229]:
# Renaming "OBS_VALUE" column in orders dataframe
pop.rename(columns = {'OBS_VALUE' : 'population'}, inplace = True)

In [230]:
pop.head()

demographic_indicator country  year  population
0                   AVG      AD  1986     45785.0
1                   AVG      AD  1987     47714.0
2                   AVG      AD  1988     49490.0
3                   AVG      AD  1989     50708.0
4                   AVG      AD  1990     52697.0

## 3.3 Checking data types

I will check the data types of all columns for all dataframes to see if anything needs to be changed. 

In [231]:
car_new.dtypes

engine_type     object
country         object
year             int64
new_car        float64
dtype: object

In [232]:
car_stock.dtypes

engine_type     object
country         object
year             int64
stock_car      float64
dtype: object

In [233]:
gdp.dtypes

country     object
year         int64
gdp_eur    float64
dtype: object

In [234]:
income.dtypes

age                 object
sex                 object
income_indicator    object
currency            object
country             object
year                 int64
income               int64
dtype: object

In [235]:
pop.dtypes

demographic_indicator     object
country                   object
year                       int64
population               float64
dtype: object

It looks that the data types for all columns in all dataframes are assigned correctly. The only change I am going to make is the "population" column in the pop dataframe, as the number of people cannot be a floating number.

In [236]:
# Convert "population" column's data type to integer
pop['population'] = pop['population'].astype('int64')

In [237]:
pop.dtypes

demographic_indicator    object
country                  object
year                      int64
population                int64
dtype: object

# 4. Data consistency checks

## 4.1 Missing values

Checking how many null values there are in each of the dataframes

#### car_new dataframe

In [238]:
car_new.isnull().sum()

engine_type     0
country         0
year            0
new_car        11
dtype: int64

There are 11 missing values in the number of new cars. This is however an important information and therefore, the null values will be kept.

#### car_stock dataframe

In [239]:
car_stock.isnull().sum()

engine_type     0
country         0
year            0
stock_car      35
dtype: int64

There are 35 missing values for the number of cars in stock. This is however an important information and therefore, the null values will be kept.

#### gdp dataframe

In [240]:
gdp.isnull().sum()

country    0
year       0
gdp_eur    0
dtype: int64

No missing values in the gdp dataframe

#### income  dataframe

In [241]:
income.isnull().sum()

age                 0
sex                 0
income_indicator    0
currency            0
country             0
year                0
income              0
dtype: int64

No missing values in the income dataframe.

#### pop dataframe

In [242]:
pop.isnull().sum()

demographic_indicator    0
country                  0
year                     0
population               0
dtype: int64

No missing values in the population dataframe

## 4.2 Duplicates

#### car_new dataframe

In [243]:
car_new_dups = car_new[car_new.duplicated()]

In [244]:
car_new_dups

Empty DataFrame
Columns: [engine_type, country, year, new_car]
Index: []

There are no duplicated values in the car_new dataframe

#### car_stock dataframe

In [245]:
car_stock_dups = car_stock[car_stock.duplicated()]

In [246]:
car_stock_dups

Empty DataFrame
Columns: [engine_type, country, year, stock_car]
Index: []

There are no duplicated values in the car_stock dataframe

#### gdp dataframe

In [247]:
gdp_dups = gdp[gdp.duplicated()]

In [248]:
gdp_dups

Empty DataFrame
Columns: [country, year, gdp_eur]
Index: []

There are no duplicated values in the gdp dataframe

#### income dataframe

In [249]:
income_dups = income[income.duplicated()]

In [250]:
income_dups

Empty DataFrame
Columns: [age, sex, income_indicator, currency, country, year, income]
Index: []

There is no duplicated values in the income dataframe

#### pop dataframe

In [251]:
pop_dups = pop[pop.duplicated()]

In [252]:
pop_dups

Empty DataFrame
Columns: [demographic_indicator, country, year, population]
Index: []

There is no duplicated values in the pop dataframe

# 5. Subsetting dataframes

## 5.1 Subsetting income dataframes

The income dataframe is very detail and we do not need that level of information. Therefore, some subsetting will be necessary.

### 5.1.1. Sex column

In [253]:
income['sex'].value_counts()

F    90624
M    90624
T    90624
Name: sex, dtype: int64

In [254]:
# Creating a subset of income dataframe for total number, without the split by sex 
income_sex = income[income['sex'] == 'T']

In [255]:
income_sex['sex'].value_counts()

T    90624
Name: sex, dtype: int64

### 5.1.2. Age column

In [256]:
income_sex['age'].value_counts()

TOTAL     4300
Y50-64    4300
Y_LT65    4300
Y_LT6     4300
Y_LT16    4300
Y_GE75    4300
Y_GE65    4300
Y_GE16    4300
Y_LT75    4300
Y25-49    4300
Y16-64    4300
Y16-24    4300
Y25-54    3732
Y55-64    3732
Y_GE18    3732
Y_GE60    3732
Y18-64    3732
Y18-24    3732
Y_LT18    3732
Y_LT60    3732
Y12-17    3732
Y6-11     3732
Y6-10      568
Y11-15     568
Y65-74     568
Name: age, dtype: int64

The possible information regarding the age, that might be useful for the analysis are:

    total number - TOTAL
    Less than 18 years - Y_LT18
    18 years and over - Y_GE18
    From 18 to 64 - Y18-64
    65 years and over - Y_GE65

In [257]:
income_age = income_sex[income_sex['age'].isin(['TOTAL', 'Y_LT18', 'Y_GE18', 'Y18-64', 'Y_GE65'])]

In [258]:
income_age.head()

age sex income_indicator currency country  year  income
8600  TOTAL   T            MED_E      EUR      AL  2017    1760
8601  TOTAL   T            MED_E      EUR      AL  2018    1997
8602  TOTAL   T            MED_E      EUR      AL  2019    2231
8603  TOTAL   T            MED_E      EUR      AL  2020    2523
8604  TOTAL   T            MED_E      EUR      AT  1995   13971

In [259]:
income_age['age'].value_counts()

TOTAL     4300
Y_GE65    4300
Y18-64    3732
Y_GE18    3732
Y_LT18    3732
Name: age, dtype: int64

### 5.1.3. Income_indicator column

In [260]:
income_age['income_indicator'].value_counts()

MED_E    9898
MEI_E    9898
Name: income_indicator, dtype: int64

MED_E stands for Median equivalise net income. MEI_E stands for mean equivalised net income. Both values should be kept and renamed in the further steps.

### 5.1.4. Currency column

In [261]:
income_age['currency'].value_counts()

EUR    7178
PPS    6428
NAC    6190
Name: currency, dtype: int64

EUR stand for Euro, NAC stands for National currency, PPS stands for 
Purchasing power standard. To simplify the analysis, I will keep the values in EUR and PPS (which might bring more insights).

In [262]:
income_curr = income_age[income_age['currency'].isin(['EUR', 'PPS'])]

In [263]:
income_curr['currency'].value_counts()

EUR    7178
PPS    6428
Name: currency, dtype: int64

In [264]:
income_curr.shape

(13606, 7)

# 5.2. Population dataframe

In [265]:
pop.head()

demographic_indicator country  year  population
0                   AVG      AD  1986       45785
1                   AVG      AD  1987       47714
2                   AVG      AD  1988       49490
3                   AVG      AD  1989       50708
4                   AVG      AD  1990       52697

### 5.2.1 Demographic_indicator column

In [266]:
pop['demographic_indicator'].value_counts()

JAN                   3077
GROW                  3011
AVG                   3010
GROWRT                3008
LBIRTH                2966
GBIRTHRT              2948
DEATH                 2929
NATGROW               2927
GDEATHRT              2913
CNMIGRAT              2913
CNMIGRATRT            2911
NATGROWRT             2911
MJAN                  2636
FJAN                  2636
FDEATH                2618
MDEATH                2618
MAVG                  2557
FAVG                  2557
MLBIRTH               1999
FLBIRTH               1999
POPSHARE              1830
POPSHARE_EU27_2020    1828
NATT                   515
NATTRT                 512
POPT                   455
MIGT                   454
POPTRT                 454
MIGTRT                 453
Name: demographic_indicator, dtype: int64

There are many demographic indicators included in the population dataset. For the analisys, we need only the average population (AVG).

In [267]:
pop_avg = pop[pop['demographic_indicator'].isin(['AVG'])]

In [268]:
pop_avg['demographic_indicator'].value_counts()

AVG    3010
Name: demographic_indicator, dtype: int64

# 6. Merging dataframes

## 6.1 Merging car_new and car_stock dataframes

In [269]:
car_new.head()

engine_type country  year  new_car
0         ALT      AL  2019   3757.0
1         ALT      AL  2020   4935.0
2         ALT      AL  2021   5703.0
3         ALT      AT  2013   1285.0
4         ALT      AT  2014   2074.0

In [270]:
car_stock.head()

engine_type country  year  stock_car
0         ALT      AL  2014     4373.0
1         ALT      AL  2015     7562.0
2         ALT      AL  2016    14544.0
3         ALT      AL  2017    21409.0
4         ALT      AL  2018    28994.0

In [271]:
car_merged = car_new.merge(car_stock, on = ['engine_type', 'country', 'year'], indicator = True, how = 'outer')

In [272]:
car_merged.head()

engine_type country  year  new_car  stock_car _merge
0         ALT      AL  2019   3757.0    35881.0   both
1         ALT      AL  2020   4935.0    43675.0   both
2         ALT      AL  2021   5703.0    52846.0   both
3         ALT      AT  2013   1285.0     5972.0   both
4         ALT      AT  2014   2074.0     7931.0   both

In [273]:
# Check to see if we hava a full march or not
car_merged['_merge'].value_counts()

both          3173
right_only    1578
left_only      196
Name: _merge, dtype: int64

In [274]:
car_merged.shape

(4947, 6)

This is an outer merge to include all the data from both dataframes and therefore, there are 1'578 rows that were only in the car_stock dataframe and 196, which were only in the car_new dataframe.
After the merge, the new dataframe has 4'947 rows.

In [275]:
# Removing the "_merge" columm
car_merged = car_merged.drop(columns = ['_merge'])

In [276]:
car_merged.head()

engine_type country  year  new_car  stock_car
0         ALT      AL  2019   3757.0    35881.0
1         ALT      AL  2020   4935.0    43675.0
2         ALT      AL  2021   5703.0    52846.0
3         ALT      AT  2013   1285.0     5972.0
4         ALT      AT  2014   2074.0     7931.0

## 6.2 Merging car_merged and gdp dataframes

In [277]:
gdp.head()

country  year  gdp_eur
0      AL  2013   9625.4
1      AL  2014   9968.6
2      AL  2015  10264.1
3      AL  2016  10719.9
4      AL  2017  11559.0

In [278]:
car_gdp = car_merged.merge(gdp, on = ['country', 'year'], indicator = True, how = 'left')

In [279]:
car_gdp.head()

engine_type country  year  new_car  stock_car   gdp_eur _merge
0         ALT      AL  2019   3757.0    35881.0   13754.2   both
1         ALT      AL  2020   4935.0    43675.0   13310.4   both
2         ALT      AL  2021   5703.0    52846.0   15157.4   both
3         ALT      AT  2013   1285.0     5972.0  323910.2   both
4         ALT      AT  2014   2074.0     7931.0  333146.1   both

In [280]:
# Check to see if we hava a full march or not
car_gdp['_merge'].value_counts()

both          4275
left_only      672
right_only       0
Name: _merge, dtype: int64

In [281]:
car_gdp.shape

(4947, 7)

The newly merged dataframe contains 4'947 rows, 4'275 were found in both dataframes and 672 only in the left one. This means that for 672 rows we will not have data about the gdp. However, I choose this approach, as the data about the cars is the most important and we cannot use the inner join in this situation.

In [282]:
# Removing the "_merge" columm
car_gdp = car_gdp.drop(columns = ['_merge'])

In [283]:
car_gdp.head()

engine_type country  year  new_car  stock_car   gdp_eur
0         ALT      AL  2019   3757.0    35881.0   13754.2
1         ALT      AL  2020   4935.0    43675.0   13310.4
2         ALT      AL  2021   5703.0    52846.0   15157.4
3         ALT      AT  2013   1285.0     5972.0  323910.2
4         ALT      AT  2014   2074.0     7931.0  333146.1

## 6.3 Merging car_gdp and income dataframes

In [284]:
income.head()

age sex income_indicator currency country  year  income
0  TOTAL   F            MED_E      EUR      AL  2017    1747
1  TOTAL   F            MED_E      EUR      AL  2018    1972
2  TOTAL   F            MED_E      EUR      AL  2019    2195
3  TOTAL   F            MED_E      EUR      AL  2020    2482
4  TOTAL   F            MED_E      EUR      AT  1995   13394

In [285]:
car_gdp_inc = car_gdp.merge(income_curr, on = ['country', 'year'], indicator = True, how = 'left')

In [286]:
car_gdp_inc.head()

engine_type country  year  new_car  stock_car  gdp_eur     age sex  \
0         ALT      AL  2019   3757.0    35881.0  13754.2   TOTAL   T   
1         ALT      AL  2019   3757.0    35881.0  13754.2   TOTAL   T   
2         ALT      AL  2019   3757.0    35881.0  13754.2   TOTAL   T   
3         ALT      AL  2019   3757.0    35881.0  13754.2   TOTAL   T   
4         ALT      AL  2019   3757.0    35881.0  13754.2  Y18-64   T   

  income_indicator currency  income _merge  
0            MED_E      EUR  2231.0   both  
1            MED_E      PPS  4252.0   both  
2            MEI_E      EUR  2619.0   both  
3            MEI_E      PPS  4992.0   both  
4            MED_E      EUR  2319.0   both

In [287]:
# Check to see if we hava a full march or not
car_gdp_inc['_merge'].value_counts()

both          89030
left_only       489
right_only        0
Name: _merge, dtype: int64

In [288]:
car_gdp_inc.shape

(89519, 12)

The newly merged dataframe now contains 71'713 rows, which is a combination of 71'224 rows which exists in both dataframes and 489 which were only in the car_gdp dataframe.

In [289]:
# Removing the "_merge" columm
car_gdp_inc = car_gdp_inc.drop(columns = ['_merge'])

In [290]:
car_gdp_inc.head()

engine_type country  year  new_car  stock_car  gdp_eur     age sex  \
0         ALT      AL  2019   3757.0    35881.0  13754.2   TOTAL   T   
1         ALT      AL  2019   3757.0    35881.0  13754.2   TOTAL   T   
2         ALT      AL  2019   3757.0    35881.0  13754.2   TOTAL   T   
3         ALT      AL  2019   3757.0    35881.0  13754.2   TOTAL   T   
4         ALT      AL  2019   3757.0    35881.0  13754.2  Y18-64   T   

  income_indicator currency  income  
0            MED_E      EUR  2231.0  
1            MED_E      PPS  4252.0  
2            MEI_E      EUR  2619.0  
3            MEI_E      PPS  4992.0  
4            MED_E      EUR  2319.0

## 6.4 Merging car_gdp_inc and pop_avg dataframes

In [291]:
pop_avg.head()

demographic_indicator country  year  population
0                   AVG      AD  1986       45785
1                   AVG      AD  1987       47714
2                   AVG      AD  1988       49490
3                   AVG      AD  1989       50708
4                   AVG      AD  1990       52697

In [292]:
ev_df = car_gdp_inc.merge(pop_avg, on = ['country', 'year'], indicator = True, how = 'left')

In [293]:
ev_df.head()

engine_type country  year  new_car  stock_car  gdp_eur     age sex  \
0         ALT      AL  2019   3757.0    35881.0  13754.2   TOTAL   T   
1         ALT      AL  2019   3757.0    35881.0  13754.2   TOTAL   T   
2         ALT      AL  2019   3757.0    35881.0  13754.2   TOTAL   T   
3         ALT      AL  2019   3757.0    35881.0  13754.2   TOTAL   T   
4         ALT      AL  2019   3757.0    35881.0  13754.2  Y18-64   T   

  income_indicator currency  income demographic_indicator  population _merge  
0            MED_E      EUR  2231.0                   AVG   2854191.0   both  
1            MED_E      PPS  4252.0                   AVG   2854191.0   both  
2            MEI_E      EUR  2619.0                   AVG   2854191.0   both  
3            MEI_E      PPS  4992.0                   AVG   2854191.0   both  
4            MED_E      EUR  2319.0                   AVG   2854191.0   both

In [294]:
# Check to see if we hava a full march or not
ev_df['_merge'].value_counts()

both          89434
left_only        85
right_only        0
Name: _merge, dtype: int64

In [295]:
ev_df.shape

(89519, 14)

The final dataframe has 71'713 rows, 71'628 rows were included in both dataframes, 85 where only in the car_gdp_inc dataframe. 

In [296]:
# Removing the "_merge" columm
ev_df = ev_df.drop(columns = ['_merge'])

# 7. Deriving new variables

The dataframe contains a lot of information which is coded and not easily understandable. Therefore, I will translate the codes into meaningful names and group some of them (e.g. engine_type).

Before adding the names to the codes, I will change the codes column names so that we know which ones are codes and which ones are names.

In [297]:
# Changing column names:
ev_df.rename(columns = {'engine_type' : 'engine_type_code'}, inplace = True)

In [298]:
ev_df.rename(columns = {'country' : 'country_code'}, inplace = True)

In [299]:
ev_df.rename(columns = {'age' : 'age_code'}, inplace = True)

In [300]:
ev_df.rename(columns = {'engine_type' : 'engine_type_code'}, inplace = True)

In [301]:
ev_df.rename(columns = {'income_indicator' : 'income_indicator_code'}, inplace = True)

In [302]:
ev_df.rename(columns = {'demographic_indicator' : 'demographic_indicator_code'}, inplace = True)

Currently the "sex" column contains only totals and there is no split between genders. Therefore, I will remove this column, as it is redundant.

In [303]:
# Removing the "sex" columm
ev_df = ev_df.drop(columns = ['sex'])

In [304]:
# Checking the first five rows from the dataframe:
ev_df.head()

engine_type_code country_code  year  new_car  stock_car  gdp_eur age_code  \
0              ALT           AL  2019   3757.0    35881.0  13754.2    TOTAL   
1              ALT           AL  2019   3757.0    35881.0  13754.2    TOTAL   
2              ALT           AL  2019   3757.0    35881.0  13754.2    TOTAL   
3              ALT           AL  2019   3757.0    35881.0  13754.2    TOTAL   
4              ALT           AL  2019   3757.0    35881.0  13754.2   Y18-64   

  income_indicator_code currency  income demographic_indicator_code  \
0                 MED_E      EUR  2231.0                        AVG   
1                 MED_E      PPS  4252.0                        AVG   
2                 MEI_E      EUR  2619.0                        AVG   
3                 MEI_E      PPS  4992.0                        AVG   
4                 MED_E      EUR  2319.0                        AVG   

   population  
0   2854191.0  
1   2854191.0  
2   2854191.0  
3   2854191.0  
4   2854191.0

## 7.1 Engine_type

In the dataset, there are 17 different engine types for passenger cars. To simplify the analysis and further visualization, I will group some engine_types into several categories.

In [305]:
# Creating categories for the engine types:
Biofuel = ['BIOETH', 'BIODIE']

In [306]:
Diesel = ['DIE', 'DIE_X_HYB']

In [307]:
Electric = ['ELC']

In [308]:
Hybrid = ['ELC_PET_HYB', 'ELC_DIE_HYB']

In [309]:
Hydrogen = ['HYD_FCELL']

In [310]:
LPG = ['LPG']

In [311]:
Other = ['ALT', 'BIFUEL', 'OTH', 'GAS']

In [312]:
Petrol = ['PET', 'PET_X_HYB']

In [313]:
Plug_in_hybrid = ['ELC_PET_PI', 'ELC_DIE_PI']

In [314]:
# Creating a new “engine” column based on the “engine_type_code” column from the ev_df dataframe.
ev_df.loc[ev_df['engine_type_code'].isin(Biofuel), 'engine'] = 'Biofuel'

In [315]:
ev_df.loc[ev_df['engine_type_code'].isin(Diesel), 'engine'] = 'Diesel'

In [316]:
ev_df.loc[ev_df['engine_type_code'].isin(Electric), 'engine'] = 'Electric'

In [317]:
ev_df.loc[ev_df['engine_type_code'].isin(Hybrid), 'engine'] = 'Hybrid'

In [318]:
ev_df.loc[ev_df['engine_type_code'].isin(Hydrogen), 'engine'] = 'Hydrogen'

In [319]:
ev_df.loc[ev_df['engine_type_code'].isin(LPG), 'engine'] = 'LPG'

In [320]:
ev_df.loc[ev_df['engine_type_code'].isin(Other), 'engine'] = 'Other'

In [321]:
ev_df.loc[ev_df['engine_type_code'].isin(Petrol), 'engine'] = 'Petrol'

In [322]:
ev_df.loc[ev_df['engine_type_code'].isin(Plug_in_hybrid), 'engine'] = 'Plug-in hybrid'

## 7.2 Country

In [323]:
# Checking the country_code column to see how many entries there are:
ev_df['country_code'].value_counts(dropna = False)

FR           3720
SE           3720
CH           3720
CY           3720
AT           3720
PT           3580
FI           3560
MT           3520
HU           3520
NL           3460
LV           3320
PL           3140
ES           3120
LU           3080
LT           2900
TR           2820
HR           2760
BE           2720
NO           2698
RO           2680
IT           2660
DE           2620
EE           2540
DK           2460
IE           2260
SI           1960
CZ           1840
UK           1595
MK           1581
AL            946
ME            801
EL            580
IS            556
XK            410
EU27_2020     290
BG            280
SK            221
LI            168
RS            148
BA            125
Name: country_code, dtype: int64

In [324]:
# Creating a dictionary of country codes and countries:
countries = {
'BE' : 'Belgium', 
'BG' : 'Bulgaria', 
'CZ' : 'Czech Republic', 
'DK' : 'Denmark', 
'DE' : 'Germany', 
'EE' : 'Estonia', 
'IE' : 'Ireland', 
'EL' : 'Greece', 
'ES' : 'Spain', 
'FR' : 'France', 
'HR' : 'Croatia', 
'IT' : 'Italy', 
'CY' : 'Cyprus', 
'LV' : 'Latvia', 
'LT' : 'Lithuania', 
'LU' : 'Luxembourg', 
'HU' : 'Hungary', 
'MT' : 'Malta', 
'NL' : 'Netherlands', 
'AT' : 'Austria', 
'PL' : 'Poland', 
'PT' : 'Portugal', 
'RO' : 'Romania', 
'SI' : 'Slovenia', 
'FI' : 'Finland', 
'SE' : 'Sweden', 
'SK' : 'Slovakia', 
'IS' : 'Iceland', 
'LI' : 'Liechtenstein', 
'NO' : 'Norway', 
'CH' : 'Switzerland', 
'UK' : 'United Kingdom', 
'BA' : 'Bosnia and Herzegovina', 
'ME' : 'Montenegro', 
'AL' : 'Albania', 
'RS' : 'Serbia', 
'TR' : 'Türkiye', 
'XK' : 'Kosovo', 
'MK' : 'North Macedonia', 
'EU27_2020' : 'European Union - 27 countries (from 2020)'
}

In [325]:
countries

{'BE': 'Belgium',
 'BG': 'Bulgaria',
 'CZ': 'Czech Republic',
 'DK': 'Denmark',
 'DE': 'Germany',
 'EE': 'Estonia',
 'IE': 'Ireland',
 'EL': 'Greece',
 'ES': 'Spain',
 'FR': 'France',
 'HR': 'Croatia',
 'IT': 'Italy',
 'CY': 'Cyprus',
 'LV': 'Latvia',
 'LT': 'Lithuania',
 'LU': 'Luxembourg',
 'HU': 'Hungary',
 'MT': 'Malta',
 'NL': 'Netherlands',
 'AT': 'Austria',
 'PL': 'Poland',
 'PT': 'Portugal',
 'RO': 'Romania',
 'SI': 'Slovenia',
 'FI': 'Finland',
 'SE': 'Sweden',
 'SK': 'Slovakia',
 'IS': 'Iceland',
 'LI': 'Liechtenstein',
 'NO': 'Norway',
 'CH': 'Switzerland',
 'UK': 'United Kingdom',
 'BA': 'Bosnia and Herzegovina',
 'ME': 'Montenegro',
 'AL': 'Albania',
 'RS': 'Serbia',
 'TR': 'Türkiye',
 'XK': 'Kosovo',
 'MK': 'North Macedonia',
 'EU27_2020': 'European Union - 27 countries (from 2020)'}

In [326]:
# Creating a for-loop to assign the country value to the country_code:
result = []

for code in ev_df['country_code']:
    result.append(countries[code])

In [327]:
# Checking the result list
result

['Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Albania',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Austria',
 'Au

In [328]:
# Adding the "result" list to the dataframe:
ev_df['country'] = result

In [329]:
ev_df['country'].value_counts(dropna = False)

France                                       3720
Sweden                                       3720
Switzerland                                  3720
Cyprus                                       3720
Austria                                      3720
Portugal                                     3580
Finland                                      3560
Malta                                        3520
Hungary                                      3520
Netherlands                                  3460
Latvia                                       3320
Poland                                       3140
Spain                                        3120
Luxembourg                                   3080
Lithuania                                    2900
Türkiye                                      2820
Croatia                                      2760
Belgium                                      2720
Norway                                       2698
Romania                                      2680


The assignment of the country names went correctly.

The "European Union - 27 countries (from 2020)" value can be removed from the dataframe, as it does not bring anything.

In [330]:
# Removing "European Union - 27 countries (from 2020)" values by creating a subset excluding this value: 
ev_df = ev_df.loc[ev_df['country'] != 'European Union - 27 countries (from 2020)']

In [331]:
# Checking if the values were removed:
ev_df['country'].value_counts(dropna = False)

France                    3720
Switzerland               3720
Cyprus                    3720
Sweden                    3720
Austria                   3720
Portugal                  3580
Finland                   3560
Malta                     3520
Hungary                   3520
Netherlands               3460
Latvia                    3320
Poland                    3140
Spain                     3120
Luxembourg                3080
Lithuania                 2900
Türkiye                   2820
Croatia                   2760
Belgium                   2720
Norway                    2698
Romania                   2680
Italy                     2660
Germany                   2620
Estonia                   2540
Denmark                   2460
Ireland                   2260
Slovenia                  1960
Czech Republic            1840
United Kingdom            1595
North Macedonia           1581
Albania                    946
Montenegro                 801
Greece                     580
Iceland 

## 7.3 Age

In [332]:
# Checking the age_code column to see how many entries there are:
ev_df['age_code'].value_counts(dropna = False)

TOTAL     17748
Y18-64    17748
Y_GE18    17748
Y_GE65    17748
Y_LT18    17748
NaN         489
Name: age_code, dtype: int64

The null values in the age column should be left, as the age is only a supplementary information to the car resistration and car stock.

The age_code should be changed to be more undestandable:

    TOTAL - total
    Y18-64 - 18-64
    Y_GE18 - 18+
    Y_LT18 - 18-
    Y_GE65 - 65+

In [333]:
# Creating new column using the loc() function
ev_df.loc[ev_df['age_code'] == 'TOTAL', 'age_group'] = 'Total'

In [334]:
ev_df.loc[ev_df['age_code'] == 'Y18-64', 'age_group'] = '18-64'

In [335]:
ev_df.loc[ev_df['age_code'] == 'Y_GE18', 'age_group'] = '18+'

In [336]:
ev_df.loc[ev_df['age_code'] == 'Y_LT18', 'age_group'] = '18-'

In [337]:
ev_df.loc[ev_df['age_code'] == 'Y_GE65', 'age_group'] = '65+'

In [338]:
# Checking the result
ev_df['age_group'].value_counts(dropna=False)

Total    17748
18-64    17748
18+      17748
65+      17748
18-      17748
NaN        489
Name: age_group, dtype: int64

The results are as expected.

## 7.4 Income indicator

MED_E stands for Median equivalise net income. MEI_E stands for mean equivalised net income. I will rename both variables to median and mean for simplification.

In [339]:
# Frequency before the change
ev_df['income_indicator_code'].value_counts(dropna=False)

MED_E    44370
MEI_E    44370
NaN        489
Name: income_indicator_code, dtype: int64

In [340]:
# Changing the values
ev_df.loc[ev_df['income_indicator_code'] == 'MED_E', 'income_indicator_code'] = 'Median'

In [341]:
ev_df.loc[ev_df['income_indicator_code'] == 'MEI_E', 'income_indicator_code'] = 'Mean'

In [342]:
# Checking the results
ev_df['income_indicator_code'].value_counts(dropna=False)

Median    44370
Mean      44370
NaN         489
Name: income_indicator_code, dtype: int64

The change was done sucessfully. I will now change back the name of the column, since it is not a code anymore.

In [343]:
ev_df.rename(columns = {'income_indicator_code' : 'income_indicator'}, inplace = True)

In [344]:
# Checking the result
ev_df.head()

engine_type_code country_code  year  new_car  stock_car  gdp_eur age_code  \
0              ALT           AL  2019   3757.0    35881.0  13754.2    TOTAL   
1              ALT           AL  2019   3757.0    35881.0  13754.2    TOTAL   
2              ALT           AL  2019   3757.0    35881.0  13754.2    TOTAL   
3              ALT           AL  2019   3757.0    35881.0  13754.2    TOTAL   
4              ALT           AL  2019   3757.0    35881.0  13754.2   Y18-64   

  income_indicator currency  income demographic_indicator_code  population  \
0           Median      EUR  2231.0                        AVG   2854191.0   
1           Median      PPS  4252.0                        AVG   2854191.0   
2             Mean      EUR  2619.0                        AVG   2854191.0   
3             Mean      PPS  4992.0                        AVG   2854191.0   
4           Median      EUR  2319.0                        AVG   2854191.0   

  engine  country age_group  
0  Other  Albania     Total  
1  Other  Albania     Total  
2  Other  Albania     Total  
3  Other  Albania     Total  
4  Other  Albania     18-64

## 7.5 Coutry grouping

Even though the data comes from European statistical office, there is data for countries outside EU. I will add an additional flag to group the countries into: EU, EFTA, other European countries.

In [345]:
# Creating categories for the country group:
EU = ['Belgium', 'Bulgaria', 'Czech Republic', 'Denmark', 'Germany', 'Estonia', 'Ireland', 'Greece', 'Spain', 'France', 'Croatia', 'Italy', 'Cyprus', 'Latvia', 'Lithuania', 'Luxembourg', 'Hungary', 'Malta', 'Netherlands', 'Austria', 'Poland', 'Portugal', 'Romania', 'Slovenia', 'Finland', 'Sweden', 'Slovakia']

In [346]:
EFTA = ['Iceland', 'Liechtenstein', 'Norway', 'Switzerland']

In [347]:
other = ['United Kingdom', 'Bosnia and Herzegovina', 'Montenegro', 'Albania', 'Serbia', 'Türkiye', 'Kosovo', 'North Macedonia']

In [348]:
# Creating a new “country_group” column based on the “country” column from the ev_df dataframe.
ev_df.loc[ev_df['country'].isin(EU), 'country_group'] = 'EU'

In [349]:
ev_df.loc[ev_df['country'].isin(EFTA), 'country_group'] = 'EFTA'

In [350]:
ev_df.loc[ev_df['country'].isin(other), 'country_group'] = 'other'

# 8. Exporting dataframe

In [351]:
# Exporting the final dataframe to pickle
ev_df.to_pickle(os.path.join(path_data, 'Prepared Data', 'ev_clean.pkl'))

In [352]:
ev_df.head()

engine_type_code country_code  year  new_car  stock_car  gdp_eur age_code  \
0              ALT           AL  2019   3757.0    35881.0  13754.2    TOTAL   
1              ALT           AL  2019   3757.0    35881.0  13754.2    TOTAL   
2              ALT           AL  2019   3757.0    35881.0  13754.2    TOTAL   
3              ALT           AL  2019   3757.0    35881.0  13754.2    TOTAL   
4              ALT           AL  2019   3757.0    35881.0  13754.2   Y18-64   

  income_indicator currency  income demographic_indicator_code  population  \
0           Median      EUR  2231.0                        AVG   2854191.0   
1           Median      PPS  4252.0                        AVG   2854191.0   
2             Mean      EUR  2619.0                        AVG   2854191.0   
3             Mean      PPS  4992.0                        AVG   2854191.0   
4           Median      EUR  2319.0                        AVG   2854191.0   

  engine  country age_group country_group  
0  Other  Albania     Total         other  
1  Other  Albania     Total         other  
2  Other  Albania     Total         other  
3  Other  Albania     Total         other  
4  Other  Albania     18-64         other

In [353]:
ev_df.shape

(89229, 16)

In [354]:
ev_df['demographic_indicator_code'].value_counts(dropna=False)

AVG    89144
NaN       85
Name: demographic_indicator_code, dtype: int64